src/tss2-tcti/tcti-spi.c

```diff
@@ -0,0 +1,625 @@
+        spi_tpm_log_register_access(access, reg_number, NULL, cnt, "failed ending the transaction");
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    // Print debug information and return success
+    spi_tpm_log_register_access(access, reg_number, buffer, cnt, NULL);
+    return TSS2_RC_SUCCESS;
+}
+
+static TSS2_RC spi_tpm_write_reg(TSS2_TCTI_SPI_CONTEXT* ctx, uint32_t reg_number, const void *buffer, size_t cnt)
+{
+    TSS2_RC rc;
+    enum TCTI_SPI_REGISTER_ACCESS_TYPE access = TCTI_SPI_REGISTER_WRITE;
+    
+    // Check maximum register transfer size is 64 byte
+    assert(cnt <= 64);
+
+    // Start write transaction
+    rc = spi_tpm_start_transaction(ctx, access, cnt, reg_number);
+    if (rc != TSS2_RC_SUCCESS) {
+        spi_tpm_end_transaction(ctx);
+        spi_tpm_log_register_access(access, reg_number, buffer, cnt, "failed in transaction start");
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+    // Write register
+    rc = spi_tpm_spi_transfer(ctx, buffer, NULL, cnt);
+    if (rc != TSS2_RC_SUCCESS) {
+        spi_tpm_end_transaction(ctx);
+        spi_tpm_log_register_access(access, reg_number, buffer, cnt, "failed in transfer");
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+    // End transaction
+    rc = spi_tpm_end_transaction(ctx);
+    if (rc != TSS2_RC_SUCCESS) {
+        spi_tpm_log_register_access(access, reg_number, NULL, cnt, "failed ending the transaction");
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    // Print debug information and return success
+    spi_tpm_log_register_access(access, reg_number, buffer, cnt, NULL);
+    return TSS2_RC_SUCCESS;
+}
+
+static uint32_t spi_tpm_read_sts_reg(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    uint32_t status = 0;
+    spi_tpm_read_reg(ctx, TCTI_SPI_TPM_STS_REG, &status, sizeof(status));
+    return status;
+}
+
+static void spi_tpm_write_sts_reg(TSS2_TCTI_SPI_CONTEXT* ctx, uint32_t status)
+{
+    spi_tpm_write_reg(ctx, TCTI_SPI_TPM_STS_REG, &status, sizeof(status));
+}
+
+static uint32_t spi_tpm_get_burst_count(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    uint32_t status = spi_tpm_read_sts_reg(ctx);
+    return (status & TCTI_SPI_TPM_STS_BURST_COUNT_MASK) >> TCTI_SPI_TPM_STS_BURST_COUNT_SHIFT;
+}
+
+static uint8_t spi_tpm_read_access_reg(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    uint8_t access = 0;
+    spi_tpm_read_reg(ctx, TCTI_SPI_TPM_ACCESS_REG, &access, sizeof(access));
+    return access;
+}
+
+static void spi_tpm_write_access_reg(TSS2_TCTI_SPI_CONTEXT* ctx, uint8_t access_bit)
+{
+    // Writes to access register can set only 1 bit at a time
+    assert (!(access_bit & (access_bit - 1)));
+    spi_tpm_write_reg(ctx, TCTI_SPI_TPM_ACCESS_REG, &access_bit, sizeof(access_bit));
+}
+
+static TSS2_RC spi_tpm_claim_locality(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    uint8_t access;
+    access = spi_tpm_read_access_reg(ctx);
+    
+    // Check if locality 0 is busy
+    if (access & TCTI_SPI_TPM_ACCESS_ACTIVE_LOCALITY) {
+        LOG_ERROR("Locality 0 is busy, status: %#x", access);
```

Comment:
If we already have the active locality this is FINE, not an error.
Localities are not really a concurrency mechanism, more a privilege level thing.
Some systems initially set once in user mode locality 0 as active and keep it this way.

Reply:
So the TCTI should just check if locality 0 is *not* already active, and only then request access via `TCTI_SPI_TPM_ACCESS_REQUEST_USE`? As I understand the PTP document requesting access to an already active locality does nothing, so we could just drop the check and error and just blindly request access every time?

Reply:
I think this is a viable solution.

Reply:
The first or the second suggested one? ;)
1. check and request if not active
2. always request and do not check